### PR TITLE
MC auto: leave room for altitude error control during descent

### PR DESCRIPTION
**Pull request by bresch**
**Describe problem solved by this pull request**
Since the same parameter is used to generate the trajectory and to saturate the controller, there is no additional space for the output of the position controller once it is filled with the feedforward, letting the altitude grow uncontrolled.

The plot below was generated in SITL with a detuned vertical velocity controller and shows an extreme case. However, I've seen the same issue on real drones when it struggles to get perfect velocity tracking due to large wind gusts.
DeepinScreenshot_select-area_20211217152122

**Describe your solution**
see https://github.com/PX4/PX4-Autopilot/pull/18872#issuecomment-1006811116

**Test data / coverage**
SITL tested

### CODE REVIEW

**Comment by ThomasRigi** [comment]
Just last week I had actually thought that it would be nice to have a dedicated descent speed in auto mode, analogous to MPC_XY_CRUISE and MPC_XY_VEL_MAX.

In our use case we want to be able to descend more aggressively in offboard mode and in altitude mode than in mission (or position) mode. In position-controlled descents we see oscillations earlier than in only altitude or even more pronounced in only attitude controlled modes. So for emergency landings and other contingency situations we want to descend faster than we'd normally do for missions.

So +1 from me if you decide to go for the separate parameters option which would solve both these issues in one go. I think it's not that much work. When I was thinking about it I thought it might be enough to replace all (three) occurrences of `MPC_Z_VEL_MAX_DN` in [FlightTaskAuto](https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp) with `MPC_Z_DN_CRUISE` (or similar). I think in all other files you could leave the existing MPC_Z_VEL_MAX_DN. And maybe add a check that MPC_Z_VEL_MAX_DN > 1.x * MPC_Z_DN_CRUISE

File: src/modules/flight_mode_manager/tasks/Orbit/FlightTaskOrbit.cpp [hunk 1]
```diff
@@ -258,9 +258,9 @@ void FlightTaskOrbit::_updateTrajectoryBoundaries()
 
 	} else { // down
 		_position_smoothing.setMaxAccelerationZ(_param_mpc_acc_down_max.get());
-		_position_smoothing.setMaxVelocityZ(_param_mpc_z_vel_max_dn.get());
-		_altitude_velocity_smoothing.setMaxVel(_param_mpc_acc_down_max.get());
```
Comment (bresch) [inline src/modules/flight_mode_manager/tasks/Orbit/FlightTaskOrbit.cpp:262]: Note that this was wrong before...

**Comment by bresch** [comment]
@ThomasRigi In summary I've made the following changes:
1. create a new `MPC_Z_V_AUTO_UP` and `MPC_Z_V_AUTO_DN` to specify the desired vertical up and down velocity in auto modes
2. make sure the constraint of the velocity controller is at least 20% higher than this value in auto modes
3. `MPC_Z_V_AUTO_[UP|DN]` are enforced to not greater than `MPC_Z_VEL_MAX_[UP|DN]`

For the rest, `MPC_Z_VEL_MAX_UP` and `MPC_Z_VEL_MAX_DN` are still used for manual modes and offboard.

**Comment by bresch** [comment]
Tested in SITL, the altitude control loop is able to add some velocity setpoint on top of the feedforward:
DeepinScreenshot_select-area_20220107104327

Also, the right parameter is applied in each mode (this flight has MPC_Z_VEL_MAX_[UP|DN] = 4 and 2, and MPC_Z_V_AUTO_[UP|DN] = 2.5 and 1.0) and the drone sets the land speed properly il auto and manual modes.
DeepinScreenshot_select-area_20220107105812

**Review by ThomasRigi — COMMENTED** [review]
Looks good to me, thanks!

**Review by ThomasRigi — DISMISSED** [review]

File: src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp [hunk 2]
```diff
@@ -827,8 +827,13 @@ void FlightTaskAuto::_updateTrajConstraints()
 
 	} else { // down
 		_position_smoothing.setMaxAccelerationZ(_param_mpc_acc_down_max.get());
-		_position_smoothing.setMaxVelocityZ(_param_mpc_z_vel_max_dn.get());
+		_position_smoothing.setMaxVelocityZ(_param_mpc_z_v_auto_dn.get());
 	}
+
+	// Stretch the constraints of the velocity controller to leave some room for an additional
+	// correction required by the altitude controller
```
Comment (MaEtUgR) [inline src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp:834]:
```suggestion
	// correction required by the altitude/vertical position controller
```
Maybe more clear, not sure

**Review by MaEtUgR — APPROVED** [review]
As discussed this strategy is acceptable and I don't see any problem in the implementation itself. 👍 👍 
We might want to simplify some parts of the logic again unrelated to what was changed here.

**Pull request merged**